2022-05-24 Future-based shuffle for Ray.md

```diff
@@ -0,0 +1,108 @@
+# Motivation
+Shuffle is the core operation for large-scale data processing. The data is partitioned according to the partition key in each record through the ALL-TO-ALL connection, and sent from all mappers to all reducers. Due to the nature of its ALL-TO_ALL connections, shuffle often becomes the bottleneck of the performance, stability and scalability of the entire system. This MEP introduces a pull-based shuffle implementation using ray future which serves following goals:
+- Improve Mars-on-Ray shuffle scalability by reducing shuffle meta
+- Improve scheduling performance by using ray c++ scheduler
+- Support shuffle failover by lineage recontruction
+- Setup the foundation for push-based shuffle using ray future
+
+## Design and Architecture
+When the Mars subtask has upstream and downstream dependencies, the downstream subtask's dependence on the output of the upstream subtask is determined by the data key. The input data required for subtask execution is a context dict object from data key to data. 
+
+For non-shuffle operators, the data key is the chunk key; but for shuffle operators, the data key is a tuple of (chunk key, reducer index). 
+
+Therefore, in order to build a Shuffle DAG based on Ray future, either all the data keys output by each mapper subtask are known in advance, or the data keys are removed.
+
+### Why not using data key?
```

Comment:
Traditionally, `xEP`s add a separate section called `alternative` at the end of the doc to describe abandoned designs. Better to follow it.

Reply:
Good suggestion, Moved to `alternative` section